api/task_check_run.go

```diff
@@ -92,6 +92,9 @@ type TaskCheckDatabaseStatementAdvisePayload struct {
 	DbType    db.Type `json:"dbType,omitempty"`
 	Charset   string  `json:"charset,omitempty"`
 	Collation string  `json:"collation,omitempty"`
+
+	// Schema review special fields.
+	EnvironmentID int `json:"environmentID,omitempty"`
```

Comment:
I would pass the policy ID instead since the check does not necessarily need to know the environment.

Also, there is a subtle difference, if we use the environment id, what if the user later changes environment <-> policy binding after the task check is created, but before the task check is run?

It's more reasonable that the policy snapshot is created when the task check is created.

You may realize that passing the policy id still has a similar problem, what if the user later changes the policy content?

TBH, passing the entire policy content might be more reasonable from the code logic point of view. However, since the new task check payload is persisted every time we start a new run, it's a bit costly to persist that entire payload. Thus using Policy ID is a decent tradeoff.

Reply:
I don't think `policy ID` should be used as an identifier here. Because `policy ID` doesn't solve the problem
> there is a subtle difference, if we use the environment id, what if the user later changes environment <-> policy binding after the task check is created, but before the task check is run?
When the environment <-> policy binding is changed, the old policy is usually deleted. In this case, the corresponding policy cannot be found through the `policy ID`.

From the user's point of view, I think there are two designs to choose from:
1. Always use the latest policy for checking. That means using the `Environment ID` as the identifier here. This design assumes that the user modifies the policy to use the updated policy. So it's reasonable to use the new policy for checks that were created before the binding was modified but run after the modification.
2. Use the policy at creation time. This design requires one more read and write operation.

I prefer the first one. Because the assumptions it brings are actually reasonable at the user level. If this assumption is not acceptable, I think only `policy snapshot `can be used instead of `policy id`.

Reply:
First, "Policy" at the creation time is the desired behavior.  Mutation after creation is problematic in the review scenario, if things change, the ideal behavior is to dismiss the previous approval if exists.

If we use "Environment ID", then we have 2 mutation points.

If we use "Policy ID", then we have 1 mutation point.

So "Policy ID" is a better solution than "Environment ID", but it's still a tradeoff compared to storing the full snapshot.

I am also fine storing the full snapshot as it removes the dependency on policy ID at all, this is desirable if we extract the "schema rule check" into a separate service (which is independent of the bytebase metadata)

e.g. we provide a SQL lint service at   api.bytebase.com/sqlcheck

Reply:
OK. I think I see what you mean.

> Mutation after creation is problematic in the review scenario, if things change, the ideal behavior is to dismiss the previous approval if exists.

I would follow this behavior.
Thanks.

Reply:
As a second thought, saving the entire snapshot is still too much at this point. We want to keep our metadata as little as possible. The small metadata has several benefits:

1. Keep our own backup small.
2. Fast schema migration on our own metadata.

And because the policy snapshot is relatively big and contains duplicate info 99% of the time, so the cost is too high.

So I would go back to suggest using the "Policy ID", it's the less evil one than the "Environment ID" at least.

Sorry about the back and forth here.